src/borrow.rs

```diff
@@ -0,0 +1,195 @@
+use std::cell::UnsafeCell;
+use std::collections::hash_map::{Entry, HashMap};
+use std::ops::Deref;
+
+use ndarray::{ArrayView, ArrayViewMut, Dimension, Ix1, Ix2, IxDyn};
+use pyo3::{FromPyObject, PyAny, PyResult};
+
+use crate::array::PyArray;
+use crate::dtype::Element;
+use crate::error::NotContiguousError;
+
+thread_local! {
+    static BORROW_FLAGS: UnsafeCell<HashMap<usize, isize>> = UnsafeCell::new(HashMap::new());
+}
+
+pub struct PyReadonlyArray<'py, T, D>(&'py PyArray<T, D>);
+
+pub type PyReadonlyArray1<'py, T> = PyReadonlyArray<'py, T, Ix1>;
+
+pub type PyReadonlyArray2<'py, T> = PyReadonlyArray<'py, T, Ix2>;
+
+pub type PyReadonlyArrayDyn<'py, T> = PyReadonlyArray<'py, T, IxDyn>;
+
+impl<'py, T, D> Deref for PyReadonlyArray<'py, T, D> {
+    type Target = PyArray<T, D>;
+
+    fn deref(&self) -> &Self::Target {
+        self.0
+    }
+}
+
+impl<'py, T: Element, D: Dimension> FromPyObject<'py> for PyReadonlyArray<'py, T, D> {
+    fn extract(obj: &'py PyAny) -> PyResult<Self> {
+        let array: &'py PyArray<T, D> = obj.extract()?;
+        Ok(array.readonly())
+    }
+}
+
+impl<'py, T, D> PyReadonlyArray<'py, T, D>
+where
+    T: Element,
+    D: Dimension,
+{
+    pub(crate) fn try_new(array: &'py PyArray<T, D>) -> Option<Self> {
+        let address = array as *const PyArray<T, D> as usize;
+
+        BORROW_FLAGS.with(|borrow_flags| {
+            // SAFETY: Called on a thread local variable in a leaf function.
+            let borrow_flags = unsafe { &mut *borrow_flags.get() };
+
+            match borrow_flags.entry(address) {
+                Entry::Occupied(entry) => {
+                    let readers = entry.into_mut();
+
+                    let new_readers = readers.wrapping_add(1);
```

Comment:
I don't think this can overflow (in other words, I can't imagine any bug that can cause overflow here because it requires at least u32::MAX times operation), but if it happens, I think it's OK to `panic!` here

Reply:
We could use a `checked_add` and panic but that would only complicate the code path as the `wrapping_add` and the check for `<= 0`, i.e. either it was `-1` and there is already a writer or it overflowed and there are `isize::MAX` readers already handles both cases without additional control flow or panics with the same result that we can have only up to `isize::MAX` readers. (The code is also not really my idea, this basically std's `BorrowFlag` implementation copied over here.)

Reply:
@kngwyu Are you satisfied with the explanation?